Source/Core/DolphinQt2/FIFOPlayerWindow.cpp

```diff
@@ -0,0 +1,348 @@
+  recording_layout->addWidget(m_frame_record_count);
+  recording_group->setLayout(recording_layout);
+
+  m_button_box = new QDialogButtonBox(QDialogButtonBox::Close);
+
+  // Action Buttons
+  m_load = m_button_box->addButton(tr("Load..."), QDialogButtonBox::ActionRole);
+  m_save = m_button_box->addButton(tr("Save..."), QDialogButtonBox::ActionRole);
+  m_record = m_button_box->addButton(tr("Record"), QDialogButtonBox::ActionRole);
+
+  m_save->setEnabled(false);
+  m_record->setEnabled(false);
+
+  layout->addWidget(info_group);
+  layout->addWidget(playback_group);
+  layout->addWidget(recording_group);
+  layout->addWidget(m_button_box);
+
+  setLayout(layout);
+}
+
+void FIFOPlayerWindow::ConnectWidgets()
+{
+  connect(m_load, &QPushButton::pressed, this, &FIFOPlayerWindow::LoadRecording);
+  connect(m_save, &QPushButton::pressed, this, &FIFOPlayerWindow::SaveRecording);
+  connect(m_record, &QPushButton::pressed, this, &FIFOPlayerWindow::ToggleRecording);
+  connect(m_button_box, &QDialogButtonBox::rejected, this, &FIFOPlayerWindow::reject);
+  connect(m_early_memory_updates, &QCheckBox::toggled, this,
+          &FIFOPlayerWindow::OnEarlyMemoryUpdatesChanged);
+  connect(m_frame_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnFrameFromChanged);
+  connect(m_frame_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnFrameToChanged);
+
+  connect(m_object_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnObjectFromChanged);
+  connect(m_object_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnObjectToChanged);
+}
+
+void FIFOPlayerWindow::LoadRecording()
+{
+  QString path = QFileDialog::getOpenFileName(this, tr("Open FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  emit LoadFIFO(path);
+}
+
+void FIFOPlayerWindow::SaveRecording()
+{
+  QString path = QFileDialog::getSaveFileName(this, tr("Save FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+
+  bool result = file->Save(path.toStdString());
+
+  if (!result)
+    QMessageBox::critical(this, tr("Error"), tr("Failed to save FIFO log."));
+}
+
+void FIFOPlayerWindow::ToggleRecording()
+{
+  FifoRecorder& recorder = FifoRecorder::GetInstance();
+
+  if (!recorder.IsRecording())
+  {
+    // Start recording
+    recorder.StartRecording(m_frame_record_count->value(),
+                            [this] { QueueOnObject(this, [this] { OnRecordingDone(); }); });
+
+    m_save->setEnabled(false);
+    m_frame_range_to->setEnabled(false);
```

Comment:
These controls' labels need to be disabled as well. It might be easiest to make a shim class like `LabelledSpinBox` to take care of that (and could be used elsewhere?).

Reply:
No, that would be to specific (Other widgets such as ``QTextEdit`` etc. would also need their own class if we follow through with that). Fine with disabling the labels though.